File: crates/haira-codegen-go/src/emitter.rs
//! GoEmitter — indented Go source code writer.

use std::collections::HashSet;

/// Writes Go source code with proper indentation.
pub struct GoEmitter {
    buf: String,
    indent: usize,
    /// Track declared variables to emit `:=` only on first assignment.
    declared_vars: HashSet<String>,
}

impl GoEmitter {
    pub fn new() -> Self {
        Self {
            buf: String::with_capacity(4096),
            indent: 0,
            declared_vars: HashSet::new(),
        }
    }

    /// Check if a variable has been declared; if not, mark it as declared.
    /// Returns true if this is a new declaration (use `:=`), false if already declared (use `=`).
    pub fn declare_var(&mut self, name: &str) -> bool {
        self.declared_vars.insert(name.to_string())
    }

    /// Reset declared variables — call at the start of each function/workflow.
    pub fn reset_vars(&mut self) {
        self.declared_vars.clear();
    }

    /// Write a line with current indentation.
    pub fn line(&mut self, s: &str) {
        if s.is_empty() {
            self.buf.push('\n');
        } else {
            for _ in 0..self.indent {
                self.buf.push('\t');
            }
            self.buf.push_str(s);
            self.buf.push('\n');
        }
    }

    /// Write text without a trailing newline, with indentation.
    pub fn write(&mut self, s: &str) {
        for _ in 0..self.indent {
            self.buf.push('\t');
        }
        self.buf.push_str(s);
    }

    /// Write text without indentation or newline (append to current line).
    pub fn append(&mut self, s: &str) {
        self.buf.push_str(s);
    }

    /// Write a formatted line.
    pub fn linef(&mut self, args: std::fmt::Arguments<'_>) {
        let s = std::fmt::format(args);
        self.line(&s);
    }

    /// Increase indentation.
    pub fn indent(&mut self) {
        self.indent += 1;
    }

    /// Decrease indentation.
    pub fn dedent(&mut self) {
        self.indent = self.indent.saturating_sub(1);
    }

    /// Write an opening brace and indent.
    pub fn open_block(&mut self, prefix: &str) {
        self.line(&format!("{} {{", prefix));
        self.indent();
    }

    /// Dedent and write a closing brace.
    pub fn close_block(&mut self) {
        self.dedent();
        self.line("}");
    }

    /// Write a blank line.
    pub fn blank(&mut self) {
        self.buf.push('\n');
    }

    /// Consume the emitter and return the generated source.
    pub fn finish(self) -> String {
        self.buf
    }
}

File: compiler/internal/codegen/emitter.go
// Package codegen generates Go source code from Haira AST.
package codegen

import (
	"fmt"
	"strings"

	"github.com/haira-lang/haira/internal/ast"
	"github.com/haira-lang/haira/internal/checker"
)

// GoEmitter writes Go source code with proper indentation.
type GoEmitter struct {
	buf          strings.Builder
	indent       int
	declaredVars map[string]bool
	scopeStack   []map[string]bool // stack of block-scoped variable declarations
	typeInfo     *checker.TypeInfo
	sourceFile   string // Haira source filename (e.g., "examples/08-structs.haira")
	sourceText   string // Haira source text (for offset→line conversion)
	inTryBlock   bool   // true when emitting inside a try body (? uses panic instead of return)
}

// LookupExprType returns the inferred checker type for an expression span, or nil.
func (e *GoEmitter) LookupExprType(span ast.Span) checker.Type {
	if e.typeInfo == nil {
		return nil
	}
	return e.typeInfo.ExprTypes[span]
}

// NewEmitter creates a new GoEmitter.
func NewEmitter() *GoEmitter {
	return &GoEmitter{
		declaredVars: make(map[string]bool),
	}
}

// DeclareVar marks a variable as declared. Returns true if this is a NEW declaration (use :=).
func (e *GoEmitter) DeclareVar(name string) bool {
	if e.declaredVars[name] {
		return false
	}
	e.declaredVars[name] = true
	// Track in current scope so PopScope can remove it
	if len(e.scopeStack) > 0 {
		e.scopeStack[len(e.scopeStack)-1][name] = true
	}
	return true
}

// PushScope enters a new block scope (for, if, while).
// Variables declared after this call will be removed when PopScope is called.
func (e *GoEmitter) PushScope() {
	e.scopeStack = append(e.scopeStack, make(map[string]bool))
}

// PopScope exits the current block scope, removing variables declared within it.
// This ensures variables declared inside a for/if/while body don't leak into
// outer scopes, matching Go's block scoping rules.
func (e *GoEmitter) PopScope() {
	if len(e.scopeStack) == 0 {
		return
	}
	top := e.scopeStack[len(e.scopeStack)-1]
	for name := range top {
		delete(e.declaredVars, name)
	}
	e.scopeStack = e.scopeStack[:len(e.scopeStack)-1]
}

// ResetVars clears declared variables — call at the start of each function/workflow.
func (e *GoEmitter) ResetVars() {
	e.declaredVars = make(map[string]bool)
}

// Line writes a line with current indentation.
func (e *GoEmitter) Line(s string) {
	if s == "" {
		e.buf.WriteByte('\n')
		return
	}
	for i := 0; i < e.indent; i++ {
		e.buf.WriteByte('\t')
	}
	e.buf.WriteString(s)
	e.buf.WriteByte('\n')
}

// Linef writes a formatted line.
func (e *GoEmitter) Linef(format string, args ...any) {
	e.Line(fmt.Sprintf(format, args...))
}

// Indent increases indentation.
func (e *GoEmitter) Indent() {
	e.indent++
}

// Dedent decreases indentation.
func (e *GoEmitter) Dedent() {
	if e.indent > 0 {
		e.indent--
	}
}

// OpenBlock writes "prefix {" and indents.
func (e *GoEmitter) OpenBlock(prefix string) {
	e.Line(prefix + " {")
	e.Indent()
}

// CloseBlock dedents and writes "}".
func (e *GoEmitter) CloseBlock() {
	e.Dedent()
	e.Line("}")
}

// Blank writes an empty line.
func (e *GoEmitter) Blank() {
	e.buf.WriteByte('\n')
}

// String returns the generated source code.
func (e *GoEmitter) String() string {
	return e.buf.String()
}

// LineDirective emits a //line directive mapping back to the Haira source location.
func (e *GoEmitter) LineDirective(span ast.Span) {
	if e.sourceFile == "" || e.sourceText == "" {
		return
	}
	line, _ := offsetToLineCol(e.sourceText, span.Start)
	e.buf.WriteString(fmt.Sprintf("//line %s:%d\n", e.sourceFile, line))
}

// offsetToLineCol converts a byte offset to 1-based line and column numbers.
func offsetToLineCol(source string, offset int) (int, int) {
	if offset < 0 {
		offset = 0
	}
	if offset > len(source) {
		offset = len(source)
	}
	line := 1
	col := 1
	for i := 0; i < offset; i++ {
		if source[i] == '\n' {
			line++
			col = 1
		} else {
			col++
		}
	}
	return line, col
}
